src/extractcode/extract_directory.py

```diff
@@ -0,0 +1,131 @@
+   Hardlinks may be recreated as regular files, not as hardlinks to the original
+   file.
+
+ - Files and directories may be renamed when their name is a duplicate. And a
+   name may be considered a duplicate ignore upper and lower case mixes even
+   on case-sensitive file systems. In particular when an archive contains the
+   same file path several times, every paths will be extracted with different
+   files names, even though using a regular tool for extraction would have
+   overwritten previous paths with the last path.
+
+ - Paths may be converted to a safe ASCII alternative that is portable across
+   OSes.
+
+ - Symlinks, relative paths and absolute paths pointing outside of the archive
+   are replaced and renamed in such a way that all the extract content of an
+   archive exist under a single target extraction directory. This process
+   includes eventually creating "synthetic" or dummy paths that did not exist in
+   the original archive.
+"""
+
+
+"""
+An ExtractEvent contains data about an archive extraction progress:
+ - `source` is the location of the archive being extracted
+ - `target` is the target location where things are extracted
+ - `done` is a boolean set to True when the extraction is done (even if failed).
+ - `warnings` is a mapping of extracted paths to a list of warning messages.
+ - `errors` is a list of error messages.
+"""
+def extract_directory(location, kinds=extractcode.default_kinds, recurse=False):
+    """
+    Walk and extract any archives found at `location` (either a file or
+    directory). Extract only archives of a kind listed in the `kinds` kind tuple.
+
+    Return an iterable of ExtractEvent tuples for each extracted archive. This
+    can be used to track extraction progress:
+
+     - one event is emitted just before extracting an archive. The ExtractEvent
+       warnings and errors are empty. The `done` flag is False.
+
+     - one event is emitted right after extracting an archive. The ExtractEvent
+       warnings and errors contains warnings and errors if any. The `done` flag
+       is True.
+
+    If `recurse` is True, extract recursively archives nested inside other
+    archives If `recurse` is false, then do not extract further an already
+    extracted archive identified by the corresponding extract suffix location.
+
+    Note that while the original file system is walked top-down, breadth-first,
+    if recurse and a nested archive is found, it is extracted to full depth
+    first before resuming the file system walk.
+    """
+    abs_input = fileutils.as_posixpath(os.path.abspath(os.path.expanduser(location)))
+    original_input = fileutils.as_posixpath(location)
+    abs_input = fileutils.as_posixpath(os.path.abspath(os.path.expanduser(location)))
+    ignored = partial(ignore.is_ignored, ignores=ignore.ignores_VCS, unignores={})
+    resources = fileutils.resource_iter(abs_input, ignored=ignored)
```

Comment:
I am confused as to why you are creating a resource iterator here. This does not sound right.
A transparent extraction iterator would receive a stream or resources and yield more resources after extraction.